### DISCUSSION TITLE
Does Remix only server-render on first request?

### DISCUSSION THREAD

**Discussion by juniorforlife** [remix-run/remix#4877] [Q&A] [answered]
As I understand from reading the documents, this is how Remix roughly works
1. On first request, `loader` will provide data to the HTML rendered on the server by Remix 
2. Once Remix has served the fully-rendered document to the browser, it loads the JS bundle and build the whole page again with React. This hydration process happens behind the scene while the user is interacting with the server-returned HTML. The app then works just like a normal SPA
3. On navigations in the browser, Remix will call `loader` via `fetch` from the browser.

If the above understanding is correct then:
1. Does server-rendering process happen only once on first request and never again?
2. If I'm using REST apis for the backend then the client request for data will have to go to Remix server then the API server. Doesn't this make the app slower?
3. When the hydration process hasn't completed, is the page gonna be static and not interactive?

**Comment by machour** [comment, marked as answer]
1. You are correct, unless you reload the page, or use `<Link reloadDocument />`
2. In my experience, using a PHP API, it adds a liiiittle overhead, but it allows to filter out the data from Remix Server to only send what's needed to your client.
3. Nope, the page is already interactive (unless you're relying on javascript for your component, ie `<button onClick...>`). Using standard HTML features (a `<button>` that submits your `<form>`) might help.

To expand a bit on 2, in my case, since I'm hosting the Remix & PHP apps on the same server, I made the API listen on localhost only on http to avoid a network roundtrip & ssl overhead. It helped.
I also like to think of the API call as a database request, if that helps you seeing this in another way.

**Reply by juniorforlife** [reply]
Thank you! I'd like to expand more on the hydration process. Let's say Facebook or Twitter was using Remix. 
On the first request, I receive a full-blown document with the newsfeed and stuff then I immediately like a post or type a comment. 

1. If hydration hasn't completed, will the like action not work? 
2. How does React take over the page once hydration completed? Will the server-rendered HTML be thrown away and replaced by the React-rendered one?
3. What's gonna happen to the comment I was typing on the server-rendered text input? Will it be replaced with the client-side one and does React preserve the comment?

**Reply by kiliman** [reply]
1. If you're using `<Form>` or `<fetcher.Form>`, Remix will convert these to standard `<form>`. So when the user clicks _like_ before the page is hydrated, the browser will send a POST and Remix will re-render your page via SSR.
2. React hydration does not generally replace your existing markup. It assumes your rendered HTML matches the React component tree and will simply update handlers and internal state to match what it would look like if it had rendered the components directly.
3. As stated in 2, existing elements will not be recreated, so the text input should remain intact with the entered text. 

Some caveats about hydration. If for some reason the HTML markup doesn't match the React component tree, then React will throw away the original DOM tree and re-render. This typically happens when you have a browser extension that injects markup into the DOM that React isn't expecting.

Also to expand on point 2 of your original question about client-side API fetch vs Remix acting as a middle-man. Since Remix apps can be deployed to the edge near your users, an API call that is made server-side will be typically faster than one called directly from your user. This is because the loader fetch from the client to Remix will be fast due to the edge. And typically your hosting provider's backbone is much faster than your user's Internet connection. So server-to-server fetch will be quicker than client fetching to the API server which may be geographically distant.

```
client -- FAST --> Remix -- FAST --> API
client ------------ SLOW ----------> API
```
And as @machour mentions, a lot of APIs over-return data. By calling the API from your loader, you can then filter out data you don't need and return just the relevant data to the client reducing bandwidth usage for your users.

**Reply by machour** [reply]
To expand on 1: It totally depends on how you implement the like action, and wether or not you make JavaScript a requirement to run it.

Considering that the browser is capable of handling your like button without any JavaScript, and by using the standard HTML & HTTP features, you could forget that Javascript exists for a sec, and implement it like this:

```html
<form action="/like" method="post">
  <button type="submit" name="post_id" value="323">👍</button>
  <span>12 users like this post</span>
</form>
```

Clicking the button would submit the form, the browser would make a request to `/like`, and you'd increments likes and redirect to the same page, that would now display `13 users like this post` with fresh data.

Now if we introduce back JavaScript to **enhance** this experience, we could avoid the full page reload by blocking the form submit, and triggering a `fetch()` instead, then when the fetch settles, we'd update the `12` to `13`.

This is what Remix's [`<Form>`](https://remix.run/docs/en/v1/components/form) basically does. Before JavaScript loads, it's a normal form, and your button would work. After it loads, it becomes able to handle the submit with a fetch. Works in both cases.

You can even take that enhancement further by doing [optimistic UI](https://remix.run/docs/en/v1/guides/optimistic-ui), that is, increment `12` to `13` **before** the fetch() settles, and rollback to display an error message if anything goes wrong.

Having your app work without JavaScript + providing a better experience with it, is called Progressive Enhancement.

**Reply by juniorforlife** [reply]
@kiliman @machour thank you so much guys! Your answers taught me a lot. Cheers!

**Reply by juniorforlife** [reply]
@kiliman @machour sorry to ping you guys directly. I'm having an issue with the hydration error. I figured out that's because of  the react-player/youtube library that I'm using. There's nothing I can do about this so my question is: will the hydration error, if not possible to fix, seriously affect the performance of the app?

**Reply by kiliman** [reply]
The hydration errors are not ideal. In fact, they break support for `defer`. My understanding is that when React 18.3 is released, it should minimize these errors.

In the meantime, I have a workaround where you hydrate the `<head>` and `<body>` separately. https://github.com/kiliman/remix-hydration-fix
